state/machine.go

```diff
@@ -958,14 +930,15 @@ func (original *Machine) advanceLifecycle(life Life, force bool, maxWait time.Du
 				UnitNames: m.doc.Principals,
 			}
 		}
-		asserts = append(asserts, bson.DocElem{
-			Name: "$or", Value: []bson.D{
-				{{"principals", bson.D{{"$size", 0}}}},
-				{{"principals", bson.D{{"$exists", false}}}},
-			},
-		})
+		asserts = append(asserts, noUnitAsserts())
 
 		if life == Dead {
+			// A machine may not become Dead until it has no
+			// containers.
+			if err := m.evaulateContainersAdvanceLifecycle(); err != nil {
+				return nil, err
+			}
+			ops = append(ops, m.noContainersOp())
```

Comment:
Part of the real change to advanceLifecycle allowed machines with container to be marked dying.